mpisppy/utils/mps_reader.py

```diff
@@ -0,0 +1,111 @@
+###############################################################################
+# mpi-sppy: MPI-based Stochastic Programming in PYthon
+#
+# Copyright (c) 2025, Lawrence Livermore National Security, LLC, Alliance for
+# Sustainable Energy, LLC, The Regents of the University of California, et al.
+# All rights reserved. Please see the files COPYRIGHT.md and LICENSE.md for
+# full copyright and license information.
+###############################################################################
+# IMPORTANT: parens in variable names will become underscore (_)
+import mip   # from coin-or (pip install mip)
+import pyomo.environ as pyo
+
+def read_mps_and_create_pyomo_model(mps_path):
+    """
+    Reads an MPS file using mip and converts it into a Pyomo ConcreteModel.
+    
+    :param mps_path: Path to the MPS file.
+    :return: Pyomo ConcreteModel.
+
+    aside: Chatgpt was almost negative help with this function...
+    """
+
+    def _domain_lookup(v):
+        # given a mip var, return its Pyomo domain
+        if v.var_type == 'C':
+            if v.lb == 0.0:
+                return pyo.NonNegativeReals
+            else:
+                return pyo.Reals
+        elif v.var_type == 'B':
+            return pyo.Binary
+        elif v.var_type == 'I':
+            return pyo.Integers
+        else:
+            raise RuntimeError(f"Unknown type from coin mip {v.var_type=}")
+        # BTW: I don't know how to query the mip object for SOS sets
+        #      (maybe it transforms them?)
+    
+    # Read the MPS file and call the coin-mip model m
+    m = mip.Model(solver_name="cbc")
+    m.read(mps_path)
+
+    # Create a Pyomo model
+    model = pyo.ConcreteModel()
+
+    varDict = dict()  # coin mip var to pyomo var
+    # Add variables to Pyomo model with their bounds and domains
+    for v in m.vars:
+        vname = v.name.replace("(","_").replace(")","_")
+        varDict[v] = pyo.Var(domain=_domain_lookup(v), bounds=(v.lb, v.ub))
+        setattr(model, vname, varDict[v])
+    #print(f"{dir(v)=}")
+
+    # Add constraints
+    for c in m.constrs:
+        # Extract terms correctly from LinExpr
+        body = sum(coeff * varDict[var] for var, coeff in c.expr.expr.items())
```

Comment:
I think in modern versions of Pyomo this creates a linear expression automatically?

Reply:
That is what I have been told.